### PR TITLE
feat(junit): add store-skipped option to emit skipped tests

### LINKED ISSUE #885: Skipped tests not shown in jUnit output when status-level="all"
The jUnit XML report does not seem to include skipped tests, even when the `status-level` is set to `"all"` in `.config/nextest.toml`.

Using the default project generated by `cargo init --lib`, I have this in `src/lib.rs`:

```rust
pub fn add(left: usize, right: usize) -> usize {
    left + right
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore]
    fn it_works() {
        let result = add(2, 2);
        assert_eq!(result, 4);
    }
}
```

...and this in `.config/nextest.toml`:

```toml
[profile.default]
status-level = "all"
final-status-level = "all"
fail-fast = false

[profile.default.junit]
path = "test-results.xml"
```

The generated XML is in `target/nextest/default/test-results.xml` as expected, but contains:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="nextest-run" tests="0" failures="0" errors="0" uuid="83992729-11dc-43e7-b5e6-58baf30be039" timestamp="2023-06-08T02:13:52.673+00:00" time="0.000">
</testsuites>
```

Skipped tests **are** reported on `stdout` though:

```none
nextest-test ⚬ cargo nextest run
    Finished test [unoptimized + debuginfo] target(s) in 0.00s
    Starting 0 tests across 1 binaries (1 skipped)
        SKIP [         ] nextest-test tests::it_works
------------
     Summary [   0.000s] 0 tests run: 0 passed, 1 skipped
        SKIP [         ] nextest-test tests::it_works
```

I had expected skipped tests to also be listed in the jUnit output.

OS: Ubuntu 22.10
Cargo: 1.70.0 (ec8a8a0ca 2023-04-25)
Rustc: 1.70.0 (90c541806 2023-05-31)
Nextest: 0.9.53 (installed via `cargo install` for dev, pre-built binary for CI)

### PR DESCRIPTION
**Pull request by liangfu**
Adds a new `junit.store-skipped` profile configuration setting that controls whether skipped tests (ignored via `#[ignore]` or filtered out) are emitted in the JUnit XML report as `<testcase>` elements with a `<skipped>` child.

Per the maintainer decision on #885, machine-readable output must stay stable by default, so the setting defaults to `false` (current behavior: skipped tests are omitted). Users opt in by setting it to `true`.

- Wire the setting through JunitImpl/JunitSettings/JunitConfig and the default profile resolution.
- Emit a skipped testcase in the aggregator when the option is enabled.
- Regenerate the repo-config JSON schema.
- Update default-config.toml, repo-config-reference.md, the JUnit docs, and the changelog.
- Add config unit tests for the default and opt-in.

Closes #885

### CODE REVIEW

**Comment by liangfu** [comment]
@sunshowers the PR is ready for review.

File: nextest-runner/src/config/elements/junit.rs [hunk 1]
```diff
@@ -70,10 +72,34 @@ impl<'cfg> JunitConfig<'cfg> {
         self.store_failure_output
     }
 
+    /// Returns true if skipped tests should be emitted as `<testcase>`
+    /// elements with a `<skipped>` child.
+    pub fn store_skipped(&self) -> bool {
+        self.store_skipped
+    }
```
Comment (sunshowers) [inline nextest-runner/src/config/elements/junit.rs]: Thinking about it a bit, I think `store_skipped` should be an enum:

```rust
enum ReportSkipPolicy {
    None,
    Ignored,
    All,
}
```

* `None` is the current behavior here.
* `Ignored` is `MismatchReason::Ignored` only.
* `All` means all variants except `NotBenchmark` (and worth adding a note somewhere that `All` will mean that partitioned runs will have duplicates).

I think we should not report the `NotABenchmark` reason in JUnit reports.

File: nextest-runner/src/reporter/aggregator/junit.rs [hunk 2]
```diff
@@ -241,17 +241,26 @@ impl<'cfg> MetadataJunit<'cfg> {
...
-                // testsuite.add_testcase(testcase);
+                    let mut testcase_status = TestCaseStatus::skipped();
+                    testcase_status.set_message(format!("Skipped: {reason}"));
+
+                    let mut testcase =
+                        TestCase::new(test_instance.test_name.as_str(), testcase_status);
+                    testcase.set_classname(test_instance.binary_id.as_str());
+
+                    testsuite.add_test_case(testcase);
+                }
```
Comment (sunshowers) [inline nextest-runner/src/reporter/aggregator/junit.rs:267]: There are now a few different implementations of skip policies:

1. In the dispatcher: https://github.com/nextest-rs/nextest/blob/797fd1db3294a45d79488e86c8b00a8d984311f0/nextest-runner/src/runner/dispatcher.rs#L849
2. In libtest: https://github.com/nextest-rs/nextest/blob/797fd1db3294a45d79488e86c8b00a8d984311f0/nextest-runner/src/reporter/structured/libtest.rs#L259-L262
3. Here.

Can we centralize this into a single spot? With the `ReportSkipPolicy` enum:

* 1 would map to `All`.
* 2 would map to `Ignored`.
* 3 would map to whatever is specified.

File: nextest-runner/src/config/elements/junit.rs [hunk 3]
```diff
@@ -109,6 +137,9 @@ impl DefaultJunitImpl {
             store_failure_output: data
                 .store_failure_output
                 .expect("junit.store-failure-output present in default profile"),
+            store_skipped: data
+                .store_skipped
+                .expect("junit.store-skipped present in default profile"),
```
Comment (sunshowers) [inline nextest-runner/src/config/elements/junit.rs]: `store_skipped` should be in the `DeserializedJunitOutput` in `config/overrides/imp.rs` so that per-test settings apply to it.

File: nextest-runner/src/reporter/aggregator/junit.rs [hunk 4]
```diff
@@ -845,4 +856,96 @@ mod tests {
...
+            kind: TestEventKind::TestSkipped {
+                stress_index: None,
+                test_instance,
+                reason: MismatchReason::Ignored,
+            },
+        })
+    }
+
+    #[test]
+    fn test_skipped_store_skipped_true() {
```
Comment (sunshowers) [inline nextest-runner/src/reporter/aggregator/junit.rs]: Thanks for writing these tests! We also need to add an integration test that validates against the actual JUnit output generated by nextest.

File: nextest-runner/repo-config-reference.md [hunk 5]
```diff
@@ -353,6 +353,13 @@ When `max-fail` is exceeded:
 - **Documentation**: [_JUnit support_](../machine-readable/junit.md)
 - **Default**: `true`
 
+#### `profile.<name>.junit.store-skipped`
```
Comment (sunshowers) [inline nextest-runner/repo-config-reference.md]: Please add `<!-- md:version 0.9.141 -->` following the pattern elsewhere in this file.

File: nextest-runner/src/reporter/aggregator/junit.rs [hunk 6]
```diff
@@ -241,17 +241,26 @@ impl<'cfg> MetadataJunit<'cfg> {
...
 
-                // let testsuite = self.testsuite_for(test_instance);
-                //
-                // let mut testcase_status = TestcaseStatus::skipped();
-                // testcase_status.set_message(format!("Skipped: {}", reason));
-                // let testcase = Testcase::new(test_instance.name, testcase_status);
-                //
-                // testsuite.add_testcase(testcase);
+                    let mut testcase_status = TestCaseStatus::skipped();
+                    testcase_status.set_message(format!("Skipped: {reason}"));
```
Comment (sunshowers) [inline nextest-runner/src/reporter/aggregator/junit.rs]: Looking at the [`fmt::Display` impl for `MismatchReason`](https://github.com/nextest-rs/nextest/blob/797fd1db3294a45d79488e86c8b00a8d984311f0/nextest-metadata/src/test_list.rs#L1005-L1021), this should be

```suggestion
                    testcase_status.set_message(format!("Skipped: test {reason}"));
```

File: nextest-runner/src/reporter/aggregator/junit.rs [hunk 7]
```diff
@@ -241,17 +241,26 @@ impl<'cfg> MetadataJunit<'cfg> {
...
-                // TODO: report skipped tests? causes issues if we want to aggregate runs across
-                // skipped and non-skipped tests. Probably needs to be made configurable.
+            TestEventKind::TestSkipped {
+                stress_index,
+                test_instance,
+                reason,
+            } => {
+                // Only emit skipped tests if the user opted in via
+                // `store-skipped`. This keeps machine-readable output stable by
+                // default.
```
Comment (sunshowers) [inline nextest-runner/src/reporter/aggregator/junit.rs]: I'd be clearer here that this is for backwards compatibility of machine-readable output.

File: nextest-runner/src/reporter/aggregator/junit.rs [hunk 8]
```diff
@@ -241,17 +241,26 @@ impl<'cfg> MetadataJunit<'cfg> {
...
-                // testcase_status.set_message(format!("Skipped: {}", reason));
-                // let testcase = Testcase::new(test_instance.name, testcase_status);
-                //
-                // testsuite.add_testcase(testcase);
+                    let mut testcase_status = TestCaseStatus::skipped();
+                    testcase_status.set_message(format!("Skipped: {reason}"));
+
+                    let mut testcase =
+                        TestCase::new(test_instance.test_name.as_str(), testcase_status);
+                    testcase.set_classname(test_instance.binary_id.as_str());
```
Comment (sunshowers) [inline nextest-runner/src/reporter/aggregator/junit.rs:262]: I'd also do `set_time(Duration::ZERO)` here.

**Review by sunshowers — COMMENTED** [review]
Thanks for doing this! The overall direction looks good -- I have a few comments, though.

**Comment by sunshowers** [comment]
I landed https://github.com/nextest-rs/nextest/pull/3479 which updates quick-junit on main to improve how skipped tests are done in JUnit a bit. Please merge/rebase so you pick up the change -- thanks!

File: nextest-runner/src/config/elements/junit.rs (hunk 1 above)
Comment (liangfu) [reply nextest-runner/src/config/elements/junit.rs]: Done. Added `ReportSkipPolicy { None, Ignored, All }` in `config/elements/junit.rs`.

File: nextest-runner/src/reporter/aggregator/junit.rs (hunk 2 above)
Comment (liangfu) [reply nextest-runner/src/reporter/aggregator/junit.rs:267]: Centralized into `ReportSkipPolicy::should_report (reason)`.

File: nextest-runner/src/config/elements/junit.rs (hunk 3 above)
Comment (liangfu) [reply nextest-runner/src/config/elements/junit.rs]: Done. Added `store_skipped: Option<ReportSkipPolicy>` to `DeserializedJunitOutput` (overrides/imp.rs).

File: nextest-runner/src/reporter/aggregator/junit.rs (hunk 4 above)
Comment (liangfu) [reply nextest-runner/src/reporter/aggregator/junit.rs]: Added `test_run_junit_store_skipped` (integration-tests/integration/main.rs).

File: nextest-runner/repo-config-reference.md (hunk 5 above)
Comment (liangfu) [reply nextest-runner/repo-config-reference.md]: Added, in both `repo-config-reference.md` and `site/src/docs/machine-readable/junit.md`.

File: nextest-runner/src/reporter/aggregator/junit.rs (hunk 7 above)
Comment (liangfu) [reply nextest-runner/src/reporter/aggregator/junit.rs]: Reworded the comment to state that skiped tests are omitted by default to keep the machine-readable output stable, and are only emitted for the reasons selected by the configured `store-skipped` policy.

File: nextest-runner/src/reporter/aggregator/junit.rs (hunk 8 above)
Comment (liangfu) [reply nextest-runner/src/reporter/aggregator/junit.rs:262]: Done — `testcase.set_time(Duration::ZERO)` with a comment noting skipped tests never execute (junit.rs).

**Comment by liangfu** [comment]
@sunshowers thanks for https://github.com/nextest-rs/quick-junit/pull/606 and the review comments. 

I rebased and addressed review comments, ready for review.

File: nextest-runner/src/config/elements/junit.rs [hunk 9]
```diff
@@ -138,7 +214,83 @@ pub(in crate::config) struct JunitImpl {
     /// Whether to store failed test output in the JUnit XML report.
     #[serde(default)]
     pub(in crate::config) store_failure_output: Option<bool>,
+    /// Which skipped tests to emit as `<testcase>` elements with a `<skipped>`
+    /// child in the JUnit XML report.
+    ///
+    /// The TOML key remains `store-skipped` for backwards compatibility with
+    /// the original boolean form. Serde parses the value as [`ReportSkipPolicy`]
+    /// (`"none"`/`"ignored"`/`"all"`) rather than a bool.
```
Comment (sunshowers) [inline nextest-runner/src/config/elements/junit.rs]: ? this comment doesn't seem relevant.

File: nextest-runner/src/config/overrides/imp.rs [hunk 10]
```diff
@@ -243,6 +244,11 @@ impl<'p> TestSettings<'p> {
         self.junit_store_failure_output.0
     }
 
+    /// Returns which skipped tests should be reported in JUnit.
+    pub fn junit_report_skipped(&self) -> ReportSkipPolicy {
+        self.junit_report_skipped.0
+    }
```
Comment (sunshowers) [inline nextest-runner/src/config/overrides/imp.rs:250]: Is this actually used? Please ensure it's used and extend the integration test accordingly.

File: site/src/docs/machine-readable/junit.md [hunk 11]
```diff
@@ -49,6 +52,7 @@ path = "junit.xml"
 # These are the default values, specified for clarity.
 store-success-output = false
 store-failure-output = true
+store-skipped = false
```
Comment (sunshowers) [inline site/src/docs/machine-readable/junit.md]: Needs an update.

File: site/src/docs/machine-readable/junit.md [hunk 12]
```diff
@@ -36,6 +36,9 @@ Configuration options supported for JUnit reports, within the `junit` section:
 `store-failure-output`
 : Whether to store output for failing tests in the `<system-out>` and `<system-err>` elements. Defaults to true.
 
+`store-skipped` <!-- md:version 0.9.141 -->
+: Which skipped tests to emit as `<testcase>` elements with a `<skipped>` child. `"none"` (the default) emits no skipped tests, keeping machine-readable output stable. `"ignored"` emits only tests skipped because they are ignored (via `#[ignore]`). `"all"` emits all skipped tests except tests skipped because they are not benchmarks. Note that with [partitioned](../running.md#partitioning-test-runs-across-multiple-cis) runs, `"all"` causes each partition's report to include the tests skipped in that partition, so merging reports produces duplicate skipped entries.
```
Comment (sunshowers) [inline site/src/docs/machine-readable/junit.md]: `../running.md#partitioning-test-runs-across-multiple-cis` is not a real anchor.

File: nextest-runner/src/config/elements/junit.rs [hunk 13]
```diff
@@ -2,8 +2,53 @@
...
+/// Controls which skipped tests are emitted as `<testcase>` elements with a
+/// `<skipped>` child in JUnit XML output.
+///
+/// Skipped tests are not emitted by default to keep machine-readable output
+/// stable. This setting opts in to reporting some or all of them.
+#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
+#[cfg_attr(feature = "config-schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "kebab-case")]
+#[cfg_attr(test, derive(test_strategy::Arbitrary))]
+pub enum ReportSkipPolicy {
```
Comment (sunshowers) [inline nextest-runner/src/config/elements/junit.rs]: `ReportSkipPolicy` is not actually JUnit-specific any more, so it should live in its own file under `elements` I think.

File: nextest-runner/src/config/elements/junit.rs [hunk 14]
```diff
@@ -2,8 +2,53 @@
...
+    /// Returns true if a test skipped for the given reason should be reported
+    /// as a skipped test under this policy.
+    ///
+    /// Tests skipped because they are not benchmarks
+    /// ([`MismatchReason::NotBenchmark`]) are never reported.
+    pub fn should_report(self, reason: MismatchReason) -> bool {
+        match self {
+            ReportSkipPolicy::None => false,
+            ReportSkipPolicy::Ignored => matches!(reason, MismatchReason::Ignored),
+            ReportSkipPolicy::All => !matches!(reason, MismatchReason::NotBenchmark),
```
Comment (sunshowers) [inline nextest-runner/src/config/elements/junit.rs]: Please use exhaustive matching here, so that there's a compile error if a new mismatch reason is added in the future.

File: integration-tests/tests/integration/main.rs [hunk 15]
```diff
@@ -900,6 +900,79 @@ fn test_run() {
...
+        .args([
+            "--manifest-path",
+            p.manifest_path().as_str(),
+            "run",
+            "-P",
+            "with-junit-skipped",
+            "-E",
+            "test(=test_success) | test(=test_ignored)",
+        ])
+        .unchecked(true)
```
Comment (sunshowers) [inline integration-tests/tests/integration/main.rs]: This should succeed, so no reason for `unchecked(true)` any more.

File: integration-tests/tests/integration/main.rs [hunk 16]
```diff
@@ -900,6 +900,79 @@ fn test_run() {
...
+        found_skipped,
+        "expected a skipped <testcase> for test_ignored in JUnit output\n{junit_xml}"
+    );
+
+    // Sanity check: nextest reported the run happened.
+    let stderr = output.stderr_as_str();
+    assert!(
+        stderr.contains("skipped") || stderr.contains("Summary"),
+        "run produced a summary, stderr: {output}"
+    );
```
Comment (sunshowers) [inline integration-tests/tests/integration/main.rs]: This is too weak -- we have a pretty great fixture model which can be used here to determine that all the right tests passed.

File: integration-tests/tests/integration/main.rs [hunk 17]
```diff
@@ -900,6 +900,79 @@ fn test_run() {
...
+        ])
+        .unchecked(true)
+        .output();
+
+    let junit_path = p.junit_path("with-junit-skipped");
+    let junit_xml = std::fs::read_to_string(&junit_path)
+        .unwrap_or_else(|e| panic!("failed to read JUnit XML from {junit_path}: {e}"));
+    let report = quick_junit::Report::deserialize_from_str(&junit_xml).unwrap_or_else(|e| {
+        panic!("failed to parse JUnit XML from {junit_path}: {e}\n{junit_xml}")
+    });
```
Comment (sunshowers) [inline integration-tests/tests/integration/main.rs]: Please extend the way the fixtures check the JUnit output (i.e. adding a notion of which tests should appear as skipped) over doing this ad-hoc test.

**Comment by sunshowers** [comment]
btw if you're using an LLM for this, please have it review your work. Claude caught most of the things I didn't.

**Comment by liangfu** [comment]
Thanks for the taking a look.

I made a few changes:
- moved `ReportSkipPolicy` into its own `report_skip_policy.rs` with explicit per-variant matching + an `ALL_VARIANTS` guard test; 
- made the skip policy per-test by threading a new junit_report_skipped field through the event/record/replay/summary pipeline so `[[overrides]]` actually apply; 
- rewrote the integration tests to use the fixture model (`junit_skipped` tracking + `verify_junit_skipped`, dropped `unchecked(true)`, added a per-test-override test);

File: site/src/docs/machine-readable/junit.md (hunk 11 above)
Comment (liangfu) [reply site/src/docs/machine-readable/junit.md]: nice catch! changed to "none"

File: nextest-runner/src/config/elements/junit.rs (hunk 13 above)
Comment (liangfu) [reply nextest-runner/src/config/elements/junit.rs]: moved `ReportSkipPolicy` into its own report_skip_policy.rs with explicit per-variant matching + an `ALL_VARIANTS` guard test

File: site/src/docs/machine-readable/junit.md (hunk 12 above)
Comment (liangfu) [reply site/src/docs/machine-readable/junit.md]: fixed to `../ci-features/partitioning.md`

**Comment by sunshowers** [comment]
Looks good, thanks! I'll take it from here. There's a few tweaks I want to make post https://github.com/nextest-rs/nextest/pull/3482 and I'll take care of merging this in.

**Review by sunshowers — APPROVED** [review]
